src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -0,0 +1,148 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.cloudbees.jenkins.plugins.bitbucket.filesystem;
+
+import com.cloudbees.jenkins.plugins.bitbucket.BitbucketSCMSource;
+import com.cloudbees.jenkins.plugins.bitbucket.BranchSCMHead;
+import com.cloudbees.jenkins.plugins.bitbucket.PullRequestSCMHead;
+import com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketApi;
+import com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketApiFactory;
+import com.cloudbees.plugins.credentials.CredentialsMatchers;
+import com.cloudbees.plugins.credentials.CredentialsProvider;
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.Util;
+import hudson.model.Item;
+import hudson.model.Queue;
+import hudson.model.queue.Tasks;
+import hudson.scm.SCM;
+import hudson.security.ACL;
+import java.io.IOException;
+import jenkins.scm.api.SCMFile;
+import jenkins.scm.api.SCMFileSystem;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.mixin.ChangeRequestCheckoutStrategy;
+
+public class BitbucketSCMFileSystem extends SCMFileSystem {
+
+    private final String ref;
+    private final BitbucketApi api;
+
+    protected BitbucketSCMFileSystem(BitbucketApi api, String ref, SCMRevision rev) throws IOException {
+        super(rev);
+        this.ref = ref;
+        this.api = api;
+    }
+
+    /**
+     *
+     * @return Return timestamp of last commit or of tag if its annotated tag
+     * @throws IOException
+     */
+    @Override
+    public long lastModified() throws IOException {
+        // TODO figure out how to implement this
+        return 0L;
+    }
+
+    @NonNull
+    @Override
+    public SCMFile getRoot() {
+        return new BitbucketSCMFile(this, api, ref);
+    }
+
+    @Extension
+    public static class BuilderImpl extends SCMFileSystem.Builder {
+
+        @Override
+        public boolean supports(SCM source) {
+            //TODO: Determine supported by checking if its git/hg bitbucket scm with proper credentials and non wildcard branch
+            return false;
+        }
+
+        @Override
+        public boolean supports(SCMSource source) {
+            return source instanceof BitbucketSCMSource;
+        }
+
+        @Override
+        public SCMFileSystem build(@NonNull Item owner, @NonNull SCM scm, @CheckForNull SCMRevision rev) {
+            return null;
+        }
+        
+        private static StandardUsernamePasswordCredentials lookupScanCredentials(@CheckForNull Item context,
+                @CheckForNull String scanCredentialsId) {
+			if (Util.fixEmpty(scanCredentialsId) == null) {
+				return null;
+			} else {
```

Comment:
Are you using tabs? Tabs are evil.

Reply:
yes of course. Never use tabs, no idea how it got in here. Fixed.